src/only.ts

```diff
@@ -0,0 +1,42 @@
+import { IterableContainer } from './_types';
+import { purry } from './purry';
+
+type Only<T extends IterableContainer> = T extends
+  | []
+  | readonly [unknown, unknown, ...Array<unknown>]
+  | readonly [unknown, ...Array<unknown>, unknown]
+  | readonly [...Array<unknown>, unknown, unknown]
+  ? undefined
+  : T extends readonly [unknown]
+    ? T[0]
+    : T[0] | undefined;
```

Comment:
Typing is hard! This could either be `T[0]` or `T[number]` and each one of them gets right some cases (`[string, boolean?]`) and gets some cases wrong (`[string?, boolean?]`, `[string?, ...boolean[]`), and vice-versa.

I think the `T[number]` is a little broader and perhaps more "real-life" useful.

```suggestion
    : T[number] | undefined;
```

Here's my playground: https://www.typescriptlang.org/play?#code/PTAEBcE8AcFNQPIDsA2kA8AVUsAe5YkATAZ1ACdYBDIge1UlAFckBrJWgdyQG0BdUAB9Q-AHygAvKGx4CxMpRr00Iluy5I+AKBCg9Afmk8ADNt16AXNJz5CpHWD1OhIs4+cvFdBqrYduADTMfhpBAHQRav68fG4eesJeyow8UaGgEWFp3PxB2ZoO8Z7U3io8mfm5weqB1dFxBsFEsABmAJZIsESFlkZITAC2AEaw5ALCLM3tnUQA3FpaUHCIDFg2cqQUJcl1GvwuYpLWsnYkWgmu58VKPqkhtfnhEQCC5ORUGPmi2hdJt48ZF5vD7oL55e4FX7bW6ZV7vT4Q0TgmpIZH1K6GSatDpdK5WGS2eRbG5lfI-JyGTAmcm9KmmFxY6ZdeYLJbwACyVGgcCIaxORIASrAAMa0ci8kjgcgdADmQT+Kkq41conEUgA3lceABpUAdUCsWCQWgtaR8KzINBYHV8UTzAC+LLZoAAGgBGI6c7lddCavS6AB6+iuwqoJFgbqs-FmoF0jJx3T0ofDACYrAqUnwY3HiNiZiGw7AAMxRyXSpAyrOxsBl2UF8MAFnT0LKtYrVd0bZlWntdtZMHgLpTnq5PN9V0DwaThYArKWpbKgkNaLQUNRNNmwPH89PwwA2ZsklJdpcrtdUDfVpp53ETsCYAAWsHDoG8AHJwKBOGLWHqkKABkYNoBmgFB61gAB2edyzlUBl1Xdd9A7GsFwrBlcyZRNQGTWAAA5D1KY9UNg+DzyQJDN1ALt0KmBMez7RYB1dIsR29Xk-SvIM72kJ8X3fT9v3IX99XAJ82nIPUQLA3dYAATmg2V9FPBCLwoq9qOEUj1xom8sJwt1jAInYeC7JS4LPRDkKo4iXC0i8dMwhYZLdSNrJg6N1JsiYMLo5y02JQi3NlDzOy869HN7J0mJdBtWLHDjJ3At0SxELtlR4OzL1CmDbIsi8kqbALjLS65AoyvKspQnLNIqpK51S4j0syoJ+mGUYrI08LfOwws3QPIrbhKxIWxSZrQFakYxkozrt1xSL+2WF0Zzin0ErALjnKghqYKCE9ATCLt+Cs2a9N6-CBtbYjdqu-bDtiTcTqS+TtsXW7iKqNLKMemSU0Mi6iJ2t73L4a6YOOnz83mxjFr3Fb2O4jaetTVzygiO7Qdez6r2+pHYBTfyMxETJ0aCisMfbL6IdxH6UpMm69tM8HaJ3XGU0Kwm6cBhniLUnNmbmhjnRdCC4fHf11qnK9H2feB+K-H8-wgMSJOA0DwJTerOcUp4wkyo7prCzKHO6nCU36jnTJ1vX7s86rzJU-9vP57p5qAA

Reply:
good catch—added some tuples with optionals to the tests too